nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -1681,13 +1992,123 @@ pub fn volume_is_read_only(
     }
 }
 
+/// Replace a Region in a VolumeConstructionRequest
+///
+/// Note that UUIDs are not randomized by this step: Crucible will reject a
+/// `target_replace` call if the replacement VolumeConstructionRequest does not
+/// exactly match the original, except for a single Region difference.
+///
+/// Note that the generation number _is_ bumped in this step, otherwise
+/// `compare_vcr_for_update` will reject the update.
+fn replace_region_in_vcr(
+    vcr: &VolumeConstructionRequest,
+    old_region: SocketAddrV6,
+    new_region: SocketAddrV6,
+) -> anyhow::Result<VolumeConstructionRequest> {
+    match vcr {
+        VolumeConstructionRequest::Volume {
+            id,
+            block_size,
+            sub_volumes,
+            read_only_parent,
+        } => Ok(VolumeConstructionRequest::Volume {
+            id: *id,
+            block_size: *block_size,
+            sub_volumes: sub_volumes
+                .iter()
+                .map(|subvol| -> anyhow::Result<VolumeConstructionRequest> {
+                    replace_region_in_vcr(&subvol, old_region, new_region)
+                })
+                .collect::<anyhow::Result<Vec<VolumeConstructionRequest>>>()?,
+
+            // Only replacing R/W regions
+            read_only_parent: read_only_parent.clone(),
+        }),
+
+        VolumeConstructionRequest::Url { id, block_size, url } => {
+            Ok(VolumeConstructionRequest::Url {
+                id: *id,
+                block_size: *block_size,
+                url: url.clone(),
+            })
+        }
+
+        VolumeConstructionRequest::Region {
+            block_size,
+            blocks_per_extent,
+            extent_count,
+            opts,
+            gen,
+        } => {
+            let mut opts = opts.clone();
+
+            for target in &mut opts.target {
+                let parsed_target: SocketAddrV6 = target.parse()?;
+                if parsed_target == old_region {
+                    *target = new_region.to_string();
+                }
+            }
+
+            Ok(VolumeConstructionRequest::Region {
+                block_size: *block_size,
+                blocks_per_extent: *blocks_per_extent,
+                extent_count: *extent_count,
+                opts,
+                gen: *gen + 1,
+            })
+        }
+
+        VolumeConstructionRequest::File { id, block_size, path } => {
+            Ok(VolumeConstructionRequest::File {
+                id: *id,
+                block_size: *block_size,
+                path: path.clone(),
+            })
+        }
+    }
+}
+
+/// Find Regions in a Volume's subvolumes list whose target match the argument
+/// IP
```

Comment:
We should add to the comment that this adds to the provided Vec of `SocketAddrV6` addresses
when it finds a match.

Also, `targets` and `target` are tricky to see here, what if we called `targets` `matched_targets` or 
something that differentiates it a little better.

Reply:
agree, done in c3cc393c4